src/inmanta/references.py

```diff
@@ -0,0 +1,450 @@
+
+
+class Argument(pydantic.BaseModel):
+    """Base class for reference (resolver) arguments"""
+
+    type: str
+    """ The type of the argument. It is used a discriminator to select the correct class
+    """
+
+    name: str
+    """ The name of the argument in the reference constructor
+    """
+
+    @abc.abstractmethod
+    def get_arg_value(self, resource: "inmanta.resources.Resource") -> object:
+        """Get the value for the argument to be able to construct the reference again
+
+        :param resource: The resource on which the reference has been defined
+        """
+
+
+class LiteralArgument(Argument):
+    """Literal argument to a reference"""
+
+    type: typing.Literal["literal"] = "literal"
+    value: PrimitiveTypes
+
+    def get_arg_value(self, resource: "inmanta.resources.Resource") -> object:
+        return self.value
+
+
+class ReferenceArgument(Argument):
+    """Use the value of another reference as an argument for the reference"""
+
+    type: typing.Literal["reference"] = "reference"
+    id: uuid.UUID
+
+    def get_arg_value(self, resource: "inmanta.resources.Resource") -> object:
+        return resource.get_reference_value(self.id)
+
+
+class GetArgument(Argument):
+    """Get a value from the resource body as value"""
+
+    type: typing.Literal["get"] = "get"
+    dict_path_expression: str
+
+    def get_arg_value(self, resource: "inmanta.resources.Resource") -> object:
+        return None
+
+
+class PythonTypeArgument(Argument):
+    """Use the python type as an argument. This is mostly used for references that are generic for a
+    type, such as core::AttributeReference
+    """
+
+    type: typing.Literal["python_type"] = "python_type"
+    value: str
+
+    def get_arg_value(self, resource: "inmanta.resources.Resource") -> object:
+        return getattr(builtins, self.value)
+
+
+class ResourceArgument(Argument):
+    """This argument provides the resource itself to the reference"""
+
+    type: typing.Literal["resource"] = "resource"
+
+    def get_arg_value(self, resource: "inmanta.resources.Resource") -> object:
+        return resource
+
+
+ArgumentTypes = typing.Annotated[
+    LiteralArgument | ReferenceArgument | GetArgument | PythonTypeArgument | ResourceArgument,
+    pydantic.Field(discriminator="type"),
+]
+""" A list of all specific types of arguments. Pydantic uses this to instantiate the correct argument class
+"""
+
+
+class BaseModel(pydantic.BaseModel):
+    """A base model class for references and mutators"""
+
+    type: ReferenceType
+    args: list[ArgumentTypes]
+
+
+class ReferenceModel(BaseModel):
+    """A reference"""
+
+    id: uuid.UUID
+
+
+class MutatorModel(BaseModel):
+    """A mutator"""
+
+
+C = typing.TypeVar("C", bound="Base")
+
+
+class Base:
+    """A base class for references and mutators"""
+
+    type: typing.ClassVar[ReferenceType]
+
+    def __init__(self, **kwargs: object) -> None:
+        self._arguments: typing.Mapping[str, object] = kwargs
```

Comment:
I wonder if it wouldn't make more sense to e.g. pick up all values from `__dict__` that don't start with `_` 

I.e. this requires all arguments to be passed down, while this information is already available?